libraries/chain/controller.cpp

```diff
@@ -3999,15 +3986,58 @@ struct controller_impl {
       bsp->verify_qc(qc_proof);
    }
 
+   // Verify a Legacy block does not have a QC block extension.
+   // (It can have a finality block header extension during transition to
+   // Savanna)
+   void verify_legacy_block_exts( const signed_block_ptr& b, const block_header_state_legacy& prev ) {
+      uint32_t block_num = b->block_num();
+
+      auto block_exts = b->validate_and_extract_extensions();
+      auto qc_ext_id = quorum_certificate_extension::extension_id();
+      bool qc_extension_present = block_exts.count(qc_ext_id) != 0;
+      EOS_ASSERT( !qc_extension_present,
+                  block_validate_exception,
+                  "Legacy block #${b} includes a QC block extension",
+                  ("b", block_num) );
+
+      auto f_ext_id = finality_extension::extension_id();
+      std::optional<block_header_extension> header_ext = b->extract_header_extension(f_ext_id);
+      if (header_ext) {
+         const auto& f_ext = std::get<finality_extension>(*header_ext);
+         EOS_ASSERT( !f_ext.qc_claim.is_strong_qc,
+                     block_validate_exception,
+                     "Transition block #${b} has a strong QC claim",
+                     ("b", block_num) );
+
+         if (auto it = prev.header_exts.find(finality_extension::extension_id()); it != prev.header_exts.end()) {
+            const auto& prev_finality_ext = std::get<finality_extension>(it->second);
+            EOS_ASSERT( f_ext.qc_claim.block_num == prev_finality_ext.qc_claim.block_num,
+                        block_validate_exception,
+                        "Transition block #${b} QC claim block_num not equal to previous QC claim block_num",
+                        ("b", block_num) );
+         } else {
+            EOS_ASSERT( f_ext.qc_claim.block_num == block_num,
+                        block_validate_exception,
+                        "Transition block #${b} QC claim block_num not equal to current block_num",
+                        ("b", block_num) );
+         }
+      }
```

Comment:
If it doesn't have `header_ext`, then we should verify that the `prev` also does not have the finality extension.

Reply:
I missed the `new_finalizer_policy_diff` check last night. Adding it.